[quant] update fused_obs_fake_quant op to accept output_fake_quant argument

supriyar opened the pull request:
Summary:
Add a new attribute to the FusedMovingAvgObsFakeQuantize that controls if the Fake Quant operation should be applied at the output of a particular layer. The motivation is to give the users additional control to control the numerics of the fake_quant operators during training. It defaults to always fake quant the output (True).

Note: We will still observer the tensors as before (only the fake_quant operation is controlled using this flag)

For example
```
input model
x -> fc1 -> fc2 -> non_quantizable_op -> fc3

After fake_quant
x -> fake_quant(x) -> fc1 -> fake_quant(fc1) -> fc2 -> fake_quant(fc2) -> non_quantizable_op -> fake_quant() -> fc3 -> fake_quantize(fc3)

With output_fake_quant disabled at the output of fc2 and fc3 (since their outputs are non-quantizable)
x -> fake_quant(x) -> fc1 -> fake_quant(fc1) -> fc2 -> non_quantizable_op -> fake_quant() -> fc3
```

Test Plan: ./buck-out/gen/caffe2/test/quantization_fx\#binary.par -r test_disable_output_fake_quant

Differential Revision: D31174526

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5507,6 +5507,81 @@ def forward(self, x):
+                ],
+            }
+            prepared = torch.quantization.quantize_fx.prepare_qat_fx(model, qconfig_dict)
+            prepared.to(device)
```
nit: do we need to move again after prepare?

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5507,6 +5507,81 @@ def forward(self, x):
+            for node in prepared.graph.nodes:
+                if node.op == 'call_module' and torch.quantization.is_activation_post_process(modules[str(node.target)]):
+                    # get the users
+                    if check_all_users_not_linear(node):
```
is this what we are planning to do in practice?

supriyar replied:
this is one example (specific to linear) of how the output fake-quant can be disabled for a certain FakeQuantize module.

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5507,6 +5507,81 @@ def forward(self, x):
+
+
+            for node in prepared.graph.nodes:
+                if node.op == 'call_module' and torch.quantization.is_activation_post_process(modules[str(node.target)]):
```
nit: torch.ao.quantization

supriyar committed:
[quant] update fused_obs_fake_quant op to accept output_fake_quant argument (#65621)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/65621

Add a new attribute to the FusedMovingAvgObsFakeQuantize that controls if the Fake Quant operation should be applied at the output of a particular layer. The motivation is to give the users additional control to control the numerics of the fake_quant operators during training. It defaults to always fake quant the output (True).

Note: We will still observer the tensors as before (only the fake_quant operation is controlled using this flag)

For example
```
input model
x -> fc1 -> fc2 -> non_quantizable_op -> fc3

After fake_quant
x -> fake_quant(x) -> fc1 -> fake_quant(fc1) -> fc2 -> fake_quant(fc2) -> non_quantizable_op -> fake_quant() -> fc3 -> fake_quantize(fc3)

With output_fake_quant disabled at the output of fc2 and fc3 (since their outputs are non-quantizable)
x -> fake_quant(x) -> fc1 -> fake_quant(fc1) -> fc2 -> non_quantizable_op -> fake_quant() -> fc3
```

Test Plan: ./buck-out/gen/caffe2/test/quantization_fx\#binary.par -r test_disable_output_fake_quant

Reviewed By: jerryzh168

Differential Revision: D31174526

fbshipit-source-id: dc41fe86f783b58d48130b048f27e143e53c7b15

janeyx99 commented:
Hello, my team has been keeping track of reverts and the reasons for revert as a metric to approximate developer experience. Could you elaborate on why this PR was reverted for record-keeping? Thank you!
